core/drivers/zynqmp_csu_puf.c

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 Foundries.io Ltd.
+ *
+ * Jorge Ramirez-Ortiz <jorge@foundries.io>
+ */
+#include <drivers/zynqmp_csu.h>
+#include <drivers/zynqmp_csu_aes.h>
+#include <drivers/zynqmp_csu_puf.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <mm/core_memprot.h>
+
+#define PUF_CMD_OFFSET			0x00
+#define PUF_CFG0_OFFSET			0x04
+#define PUF_CFG1_OFFSET			0x08
+#define PUF_SHUT_OFFSET			0x0C
+#define PUF_STATUS_OFFSET		0x10
+#define PUF_WORD_OFFSET			0x18
+
+#define PUF_REGISTRATION		1
+#define PUF_REGENERATION		4
+#define PUF_STOP			3
+#define PUF_RESET			6
+
+#define PUF_CFG0_DEFAULT		0x02
+#define PUF_CFG1_DEFAULT		0x0c230090
+#define PUF_SHUT_DEFAULT		0x01000100
+#define PUF_REGEN_TIME_MS		3000
+#define PUF_STATUS_SYN_WRD_RDY_TIMEOUT	500000
+#define PUF_STATUS_SYN_WRD_RDY_MASK	BIT(0)
+#define PUF_STATUS_KEY_RDY_MASK		BIT(3)
+
+#define	PUF_SYN_WORDS			141
+
+TEE_Result zynqmp_csu_puf_regenerate(void)
+{
+	vaddr_t puf = core_mmu_get_va(ZYNQMP_CSU_PUF_BASE, MEM_AREA_IO_SEC,
+				      ZYNQMP_CSU_PUF_SIZE);
+	vaddr_t csu = core_mmu_get_va(CSU_BASE, MEM_AREA_IO_SEC, CSU_SIZE);
+	uint32_t status = 0;
+
+	if (!puf || !csu)
+		return TEE_ERROR_GENERIC;
+
+	io_write32(puf + PUF_CFG0_OFFSET, PUF_CFG0_DEFAULT);
+	io_write32(puf + PUF_SHUT_OFFSET, PUF_SHUT_DEFAULT);
+	io_write32(puf + PUF_CMD_OFFSET, PUF_REGENERATION);
```

Comment:
As this is not nicely documented in public I must resort to what Xilinx has been thinking what should in addition to be done:
- First they think it needs to be verified that some internal clocks are operating in expected speeds (probably affects the outcome of PUF ring oscillators' re-generation abilities)
- They validate that security eFuses have been programmed in expected state
- They update/configure the PUF helper data which I believe is quite important step in here

Reply:
I think you are talking about the PUF registration process - which must be done using the Xilinx tools and which sets all the efuses. Without that, the PUF will not be able to generate a persistent key.

Reply:
So securing the board - ie, writing the RSA_EN eFuse and the PPK key hash - must be followed by registering the PUF in order to be able to use this driver..